libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs

```diff
@@ -0,0 +1,154 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Web;
+using Microsoft.Bot.Builder.Core.Extensions;
+using Microsoft.WindowsAzure.Storage;
+using Microsoft.WindowsAzure.Storage.Blob;
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Builder.Azure
+{
+    /// <summary>
+    /// Models IStorage using Azure Storge Blobs
+    /// </summary>
+    /// <remarks>
+    /// The AzureBlobStorage implements State's IStorage using a single Azure Storage Blob Container.
+    /// Each entity or StoreItem is serialized into a JSON string and stored in an individual text blob.
+    /// Each blob is named after the StoreItem key which is encoded and ensure it conforms a valid blob name.
+    /// </remarks>
+    public class AzureBlobStorage : IStorage
+    {
+        private readonly static JsonSerializerSettings serializationSettings = new JsonSerializerSettings()
+        {
+            // we use all so that we get typed roundtrip out of storage, but we don't use validation because we don't know what types are valid
+            TypeNameHandling = TypeNameHandling.All
+        };
+
+        /// <summary>
+        /// The Azure Storage Blob Container where entities will be stored
+        /// </summary>
+        public Lazy<CloudBlobContainer> Container { get; private set; }
+
+        /// <summary>
+        /// Creates the AzureBlobStorage instance
+        /// </summary>
+        /// <param name="dataConnectionString">Azure Storage connection string</param>
+        /// <param name="containerName">Name of the Blob container where entities will be stored</param>
+        public AzureBlobStorage(string dataConnectionString, string containerName)
+            : this(CloudStorageAccount.Parse(dataConnectionString), containerName)
+        {
+        }
+
+        /// <summary>
+        /// Creates the AzureBlobStorage instance
+        /// </summary>
+        /// <param name="storageAccount">Azure CloudStorageAccount instance</param>
+        /// <param name="containerName">Name of the Blob container where entities will be stored</param>
+        public AzureBlobStorage(CloudStorageAccount storageAccount, string containerName)
+        {
+            if (storageAccount == null) throw new ArgumentNullException(nameof(storageAccount));
+
+            // Checks if a container name is valid
+            NameValidator.ValidateContainerName(containerName);
+
+            this.Container = new Lazy<CloudBlobContainer>(() =>
+              {
+                  var blobClient = storageAccount.CreateCloudBlobClient();
+                  var container = blobClient.GetContainerReference(containerName);
+                  container.CreateIfNotExistsAsync().Wait();
+                  return container;
+              });

```

Comment:
I would _not_ recommend using a `Lazy<T>` for this given that you're making an async call and forcing it to block with `.Wait()`. Since all the places you would ever call this from right now are themselves async, I would suggest just making a private helper method that acts as a factory around a private field. You don't really _need_ to worry about multiple threads initializing the container concurrently if you don't want to as there's no destructive side effects. The only potential downside is the extra call to `CreateIfNotExistsAsync` happening, which if you want to set up a gate to prevent you could, it just might be overkill.

```CSharp
private CloudBlobContainer _container;

public async ValueTask<CloudBlobContainer> GetContainer()
{
    if(_container == null)
    {
	var blobClient = storageAccount.CreateCloudBlobClient();
              _container = blobClient.GetContainerReference(containerName);
              
              await _container.CreateIfNotExistsAsync().ConfigureAwait(false);
    }

    return _container;
}
```

Like I said, if you truly truly cared about the possibility of multiple threads initializing the field... then you could introduce a gate object that you could `lock` on to prevent that, but I honestly think it's overkill. Note that I'm using a `ValueTask<T>` here as well which, given that all calls after the value is initialized will technically complete synchronously, will reduce the GC overhead of the call significantly vs. using a full `Task<T>`.

Reply:
We tried this approach and found that being not thread safe makes batch Write/Read operations to fail. Hence, we must add `lock`. Then, we had to `CreateIfNotExistsAsync().Wait()` again since await is not allowed within the `lock` body.
With these changes to the proposed code, we ended up with the same `Lazy<>` initialization code. We should add that, by default, `Lazy<>` initialization IS thread safe out-of-the-box.
For the time being we explicitly set Lazy's `isThreadSafe: true`. Implementing an async initialization factory for `Lazy` would be another option if we want to completely avoid a blocking call to `Lazy.Value`, which will happen only once.

Reply:
Let me know what you think about this last version of the Container factory and usage. I think we got the best of both worlds here:
```csharp
this.Container = new Lazy<ValueTask<CloudBlobContainer>>(async () =>
{
    var blobClient = storageAccount.CreateCloudBlobClient();
    var container = blobClient.GetContainerReference(containerName);
    await container.CreateIfNotExistsAsync().ConfigureAwait(false);
    return container;
}, isThreadSafe: true);
```

Usage:
```csharp
var blobName = GetBlobName(key);
var blobContainer = await this.Container.Value;
var blobReference = blobContainer.GetBlobReference(blobName);
```